spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CosineProximityCriterion.scala

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.abstractnn.TensorCriterion
+import com.intel.analytics.bigdl.tensor._
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * The negative of the mean cosine proximity between predictions and targets.
+ * The cosine proximity is defined as below:
+ * x'(i) = x(i) / sqrt(max(sum(x(i)^2), 1e-12))
+ * y'(i) = y(i) / sqrt(max(sum(x(i)^2), 1e-12))
+ * cosine_proximity(x, y) = mean(-1 * x'(i) * y'(i))
+ */
+class CosineProximityCriterion[@specialized(Float, Double) T: ClassTag]
+(implicit ev: TensorNumeric[T]) extends TensorCriterion[T] {
+
+  override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
+    val x = l2Norm(input)
+    val y = l2Norm(target)
+    val mul = x.cmul(y)
+    var loss = ev.fromType[Double](0.0)
+    val func = new TensorFunc2[T] {
+      override def apply(v1: Array[T], v2: Int): Unit = {
+        loss = ev.plus(ev.negative(v1(v2)), loss)
+      }
+    }
+    DenseTensorApply.apply1(mul, func)
+    loss = ev.divide(loss, ev.fromType(mul.nElement()))
+    loss
+  }
+
+  private def l2Norm(input : Tensor[T]) : Tensor[T] = {
```

Comment:
Looks like a good utility method. not sure if we can move it to Tensor class. or maybe we can put it into a following PR.